### PR TITLE
Report dataset metrics in monitoring log reporter

### PR DESCRIPTION
**Pull request by andrewkroh**
## What does this PR do?

This adds periodic reporting of "dataset" metrics to the logs. This can be disabled by configuring `logging.metrics.namespaces`. These are the same metrics available from the HTTP monitoring endpoint under the /dataset path. This primarily improves visibility of Filebeat and Metricbeat because they make use of the "dataset" namespace for input and metricset instance metrics.

## Why is it important?

Prior to 7.10 the Filebeat input metrics were reported in the logged metrics, but this was lost when the metrics moved to the "dataset" namespace. It was a useful feature to have for remote debugging.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates https://github.com/elastic/beats/pull/19977
- Relates https://github.com/elastic/beats/pull/6836

## Logs

```
2021-05-16T14:20:17.885-0400    INFO    [monitoring]    log/log.go:183  Non-zero metrics in the last 30s        {"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":60,"time":{"ms":60}},"total":{"ticks":250,"time":{"ms":250},"value":250},"user":{"ticks":190,"time":{"ms":190}}},"info":{"ephemeral_id":"caa6d941-b4ad-44fa-a5b2-6b07e41427d2","uptime":{"ms":30055}},"memstats":{"gc_next":35415664,"memory_alloc":17979752,"memory_sys":77415424,"memory_total":75690904,"rss":86929408},"runtime":{"goroutines":62}},"filebeat":{"events":{"active":534,"added":543,"done":9},"harvester":{"open_files":8,"running":8,"started":8}},"libbeat":{"config":{"module":{"running":0}},"output":{"events":{"active":0},"type":"elasticsearch"},"pipeline":{"clients":1,"events":{"active":528,"filtered":15,"published":528,"retry":150,"total":543},"queue":{"max_events":4096}}},"registrar":{"states":{"current":9,"update":9},"writes":{"success":9,"total":9}},"system":{"cpu":{"cores":12},"load":{"1":1.6616,"15":2.3765,"5":2.1978,"norm":{"1":0.1385,"15":0.198,"5":0.1831}}}}, "dataset": {"069dc32a-b71f-4344-a600-fa06d5071036":{"last_event_published_time":"2021-05-16T14:19:47.918Z","last_event_timestamp":"2021-05-16T14:19:47.918Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/pan_inc_traffic_ietf.log","read_offset":51624,"size":51624,"start_time":"2021-05-16T14:19:47.886Z"},"0a414085-71c7-4e62-86c8-d4ef09a2292b":{"last_event_published_time":"2021-05-16T14:19:47.888Z","last_event_timestamp":"2021-05-16T14:19:47.888Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/pan_inc_other.log","read_offset":6217,"size":6217,"start_time":"2021-05-16T14:19:47.884Z"},"12be14ea-a067-480f-9b3a-7f043a826068":{"last_event_published_time":"2021-05-16T14:19:47.886Z","last_event_timestamp":"2021-05-16T14:19:47.885Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/global_protect.log","read_offset":1641,"size":1641,"start_time":"2021-05-16T14:19:47.884Z"},"12cf5432-b0ff-4cb6-8db3-098e54b9cc4b":{"last_event_published_time":"2021-05-16T14:19:47.898Z","last_event_timestamp":"2021-05-16T14:19:47.898Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/userid.log","read_offset":2626,"size":2626,"start_time":"2021-05-16T14:19:47.890Z"},"8c19b7d1-a670-4e91-9950-6a96e0c49457":{"last_event_published_time":"2021-05-16T14:19:47.920Z","last_event_timestamp":"2021-05-16T14:19:47.920Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/traffic.log","read_offset":46385,"size":46385,"start_time":"2021-05-16T14:19:47.888Z"},"926dd152-78a7-45e8-abea-778e22794590":{"last_event_published_time":"2021-05-16T14:19:47.906Z","last_event_timestamp":"2021-05-16T14:19:47.905Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/pan_inc_traffic.log","read_offset":36298,"size":36298,"start_time":"2021-05-16T14:19:47.885Z"},"c089fcd5-1822-4017-9739-4890d65ca098":{"last_event_published_time":"2021-05-16T14:19:47.915Z","last_event_timestamp":"2021-05-16T14:19:47.915Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/threat.log","read_offset":41586,"size":41586,"start_time":"2021-05-16T14:19:47.887Z"},"ddec21fc-29fb-4ed5-b23c-2cc6f6f09d12":{"last_event_published_time":"2021-05-16T14:19:47.914Z","last_event_timestamp":"2021-05-16T14:19:47.914Z","name":"/Users/akroh/go/src/github.com/elastic/beats/x-pack/filebeat/module/panw/panos/test/pan_inc_threat.log","read_offset":39870,"size":39870,"start_time":"2021-05-16T14:19:47.884Z"}}}}
2021-05-16T14:20:47.883-0400    INFO    [monitoring]    log/log.go:183  Non-zero metrics in the last 30s        {"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":66,"time":{"ms":6}},"total":{"ticks":262,"time":{"ms":12},"value":262},"user":{"ticks":196,"time":{"ms":6}}},"info":{"ephemeral_id":"caa6d941-b4ad-44fa-a5b2-6b07e41427d2","uptime":{"ms":60054}},"memstats":{"gc_next":35415664,"memory_alloc":19664576,"memory_total":77375728,"rss":87179264},"runtime":{"goroutines":62}},"filebeat":{"harvester":{"open_files":8,"running":8}},"libbeat":{"config":{"module":{"running":0}},"output":{"events":{"active":0}},"pipeline":{"clients":1,"events":{"active":528,"retry":50}}},"registrar":{"states":{"current":9}},"system":{"load":{"1":1.6016,"15":2.3442,"5":2.1299,"norm":{"1":0.1335,"15":0.1954,"5":0.1775}}}}}}
2021-05-16T14:23:35.401-0400    INFO    [monitoring]    log/log.go:191  Total metrics   {"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":107,"time":{"ms":107}},"total":{"ticks":375,"time":{"ms":375},"value":375},"user":{"ticks":268,"time":{"ms":268}}},"info":{"ephemeral_id":"caa6d941-b4ad-44fa-a5b2-6b07e41427d2","uptime":{"ms":227570}},"memstats":{"gc_next":35912432,"memory_alloc":22875816,"memory_sys":77415424,"memory_total":86230296,"rss":92229632},"runtime":{"goroutines":10}},"filebeat":{"events":{"active":535,"added":544,"done":9},"harvester":{"closed":8,"open_files":0,"running":0,"skipped":0,"started":8},"input":{"log":{"files":{"renamed":0,"truncated":0}},"netflow":{"flows":0,"packets":{"dropped":0,"received":0}}}},"libbeat":{"config":{"module":{"running":0,"starts":0,"stops":0},"reloads":0,"scans":0},"output":{"events":{"acked":0,"active":0,"batches":0,"dropped":0,"duplicates":0,"failed":0,"toomany":0,"total":0},"read":{"bytes":0,"errors":0},"type":"elasticsearch","write":{"bytes":0,"errors":0}},"pipeline":{"clients":0,"events":{"active":529,"dropped":0,"failed":0,"filtered":15,"published":529,"retry":400,"total":544},"queue":{"acked":0,"max_events":4096}}},"registrar":{"states":{"cleanup":0,"current":9,"update":9},"writes":{"fail":0,"success":9,"total":9}},"system":{"cpu":{"cores":12},"load":{"1":3.0503,"15":2.4072,"5":2.4136,"norm":{"1":0.2542,"15":0.2006,"5":0.2011}}}}}}
```

### CODE REVIEW

**Comment by ruflin** [comment]
@simitt This might affect apm?

**Comment by simitt** [comment]
> This can be disabled by configuring `logging.metrics.namespaces`. 

@andrewkroh I haven't tested this, but does that mean that `logging.metrics.enabled: false` will no longer have the same effect to disable metrics in logs?

**Comment by andrewkroh** [comment]
Hi @simitt, `logging.metrics.enabled: false` will continue to have the same behavior in that it entirely disables the log reporter component of the monitoring framework. 

The `logging.metrics.namespaces` setting allows you to select what metric namespaces are reported. The old implicit behavior was `logging.metrics.namespaces: [stats]` where stats is the same metrics as the [http endpoint /stats](https://www.elastic.co/guide/en/beats/filebeat/master/http-endpoint.html#_stats). This adds reporting of the `dataset` namespace to that list. I think only Filebeat and Metricbeat actually make use of this (e.g. `monitoring.GetNamespace("dataset").GetRegistry()`).

**Comment by andrewkroh** [comment]
run tests

**Review by kvch — CHANGES_REQUESTED** [review]
The namespace "dataset" contains lots of metrics about every open file in Filebeat. Those metrics were moved to the dataset endpoint on purpose later because users did not want to see that much information about their files. I am all for this new option, but please do not add "dataset" to the default namespaces.

File: libbeat/monitoring/report/log/log.go [hunk 1]
```diff
@@ -63,6 +64,18 @@ var gauges = map[string]bool{
 	"system.load.norm.15":            true,
 }
 
+// isGauge returns true when the given metric key name represents a gauge value.
+// Any metric name suffixed in '_gauge' or containing '.histogram.' is
+// treated as a gauge. Other metrics can specifically be marked as gauges
+// through the list maintained in this package.
+func isGauge(key string) bool {
```
Comment (kvch) [inline libbeat/monitoring/report/log/log.go:71]: Where do you use this function?

Comment (andrewkroh) [reply libbeat/monitoring/report/log/log.go:71]: Fixed it and added a test to verify it.

**Review by kvch — APPROVED** [review]

**Pull request merged**